### PR TITLE
Fix matrix multiplication of rectangular ZeroLazyTensor

### PR DESCRIPTION
**Pull request by PFMassiani**
Fixes https://github.com/cornellius-gp/gpytorch/issues/1294.
The code given in the bug report to reproduce the issue is also added as a unit test.

I bring your attention to the returns of `ZeroLazyTensor._matmul` and 'ZeroLazyTensor._t_matmul': a new `torch.tensor` is created, and therefore all the gradient information that may be present in the parameter `rhs` is lost. I do not think this is a problem (backpropagation throught a `ZeroLazyTensor` does not make a lot of sense), but this should definitely be double-checked.

### CODE REVIEW

**Comment by jacobrgardner** [comment]
I'll merge this one in after the previous PR is merged and we rebase this on top of that. I think the backpropagation is a non-issue.

**Technically**, multiplying a tensor that requires grad by 0 results in a tensor with a `grad_fn`:

```python
In [1]: import torch

In [2]: a = torch.randn(5).requires_grad_(True)

In [3]: b = 0 * a

In [4]: b
Out[4]: tensor([-0., 0., -0., 0., 0.], grad_fn=<MulBackward0>)
```
This could result in insidious bugs where we rely on simply checking whether a tensor requires grad or not to clear a cache or something?

File: gpytorch/lazy/zero_lazy_tensor.py
Comment (jacobrgardner) [inline gpytorch/lazy/zero_lazy_tensor.py:53]:
```suggestion
        return torch.zeros(*output_shape, device=rhs.device, dtype=rhs.dtype)
```
At a minimum, these should inherit

File: gpytorch/lazy/zero_lazy_tensor.py
Comment (jacobrgardner) [inline gpytorch/lazy/zero_lazy_tensor.py:91]:
```suggestion
        return torch.zeros(*output_shape, dtype=rhs.dtype, device=rhs.device)
```
ditto here

**Comment by PFMassiani** [comment]
True: the device and dtype should be preserved now. I think you forgot to point it in the standard `matmul` method: this is also fixed.

About the gradient: the `requires_grad` property of ZeroLazyTensors seems to be always False. Therefore, if this variable is correctly checked before using gradient information, it should not be a problem?
```python
from gpytorch.lazy import ZeroLazyTensor
lz = ZeroLazyTensor(5, 4, 3)
print(f'Requires grad: {lz.requires_grad}')
lz.requires_grad_(True)
print(f'Requires grad: {lz.requires_grad}')
```
Output:
```
Requires grad: False
Requires grad: False
```

**Comment by jacobrgardner** [comment]
Not sure why the stable build failed the first time -- the error message made no sense. Hopefully it will pass this time.

**Review by gpleiss — APPROVED** [review]

**Pull request merged**
